diag_manager/include/fms_diag_fieldbuff_update.fh

```diff
@@ -0,0 +1,1336 @@
+
+    !!Note that diag manager does not allocate the ofc in all situations
+    if(allocated(ofc)) then
+      ofc_ptr(1:size(ofc,1),1:size(ofc,2),1:size(ofc,3), 1:1, 1:size(ofc,4)) => ofc
+    else
+      allocate(ofc_dummy(1,1,1,1))
+      ofc_ptr(1:1,1:1,1:1, 1:1,1:1) => ofc_dummy
+    endif
+
+    IF  (PRESENT (mask)) THEN
+      mask_ptr(1:size(mask,1),1:size(mask,2),1:size(mask,3),1:1) => mask
+    ENDIF
+
+    succeded = FMS_DIAG_FBU_PNAME_ (ofield_cfg, ofield_index_cfg, field_ptr, sample, &
+      & ofb_ptr, ofc_ptr, bbounds, count_0d, num_elements, mask_ptr, weight1, missvalue,  &
+      & field_num_threads, field_active_omp_level, issued_mask_ignore_warning, &
+      & l_start, l_end, err_msg,  err_msg_local)
+  END FUNCTION FMS_DIAG_FBU_3D_PNAME_
+
+
+!> @brief This code will be used by the preprecessor to generate an implementation
+!! to the module procudure for the fieldbuff_update interface.
+!! Updates elements of the running field output buffer (argument ofb)
+!! and counter (argument ofc) based on the input field data array (argument field_d).
+!! In general the formulas are :
+!! A) ofb(l) = ofb(l) + (weight * field(l))**pow_value
+!! B) ofc(l) = ofc(l) + weight
+!! where l is a standing for some set of indecies in multiple dimensions.
+!! Note this function may set field object members active_omp_level and  num_threads.
+!> @addtogroup fms_diag_fieldbuff_update_mod
+!> @{
+   FUNCTION FMS_DIAG_FBU_PNAME_ (ofield_cfg, ofield_index_cfg, field_d, sample, &
+    & ofb, ofc, bbounds, count_0d, num_elements, mask, weight1, missvalue,  &
+    & field_num_threads, field_active_omp_level, issued_mask_ignore_warning, &
+    & l_start, l_end, err_msg,  err_msg_local ) result( succeded )
+      TYPE(fms_diag_outfield_type), INTENT(in) :: ofield_cfg  !< The ofield_cfg object
+      TYPE(fms_diag_outfield_index_type) , INTENT(in) :: ofield_index_cfg  !< The ofield_index_cfg object
+      FMS_DIAG_FBU_DATA_TYPE_ , CONTIGUOUS, DIMENSION(:,:,:,:), INTENT(in) :: field_d !< The input field data array.
+      FMS_DIAG_FBU_DATA_TYPE_ , DIMENSION(:,:,:,:,:), INTENT(inout) :: ofb !< Output Field Buffer
+      FMS_DIAG_FBU_DATA_TYPE_ , DIMENSION(:,:,:,:,:), INTENT(inout) :: ofc !< Output Field Counter
+      TYPE(fms_diag_ibounds_type), INTENT(inout) :: bbounds !< An instance of the class fms_diag_buff_intervals_t
+      INTEGER, INTENT(in) :: sample  !< The index along the diurnal time axis
+      FMS_DIAG_FBU_DATA_TYPE_, INTENT(inout):: count_0d !< Normally the member of the buffer object of the same name.
+      INTEGER, INTENT(inout) :: num_elements
+
+      LOGICAL, CONTIGUOUS, DIMENSION(:,:,:,:), INTENT(in), OPTIONAL :: mask !< mask
+      FMS_DIAG_FBU_DATA_TYPE_, INTENT(in) :: weight1 !< Field data is multiplied by weight
+      FMS_DIAG_FBU_DATA_TYPE_, INTENT(in) :: missvalue !< Buffer may be set to missvalue where mask is false.
+
+      INTEGER, INTENT(inout) :: field_num_threads
+      INTEGER, INTENT(inout) :: field_active_omp_level
```

Comment:
Are these things that really need to be passed in?  I suppose it makes some sense to do them at a higher level.  This might need to be revisited in the modern implementation

Reply:
Adding this to a todo list, plus added this MDM comment in the code
!! TODO: (MDM) revisit passing in and need of field_num_threads and field_active_omp_level